File: plugins/textmate/core/src/org/jetbrains/plugins/textmate/language/preferences/SnippetsRegistryImpl.java
package org.jetbrains.plugins.textmate.language.preferences;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.plugins.textmate.language.TextMateScopeComparator;
import org.jetbrains.plugins.textmate.language.syntax.lexer.TextMateScope;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public final class SnippetsRegistryImpl implements SnippetsRegistry {
  private final @NotNull Map<String, Collection<TextMateSnippet>> mySnippets = new ConcurrentHashMap<>();

  public void register(@NotNull TextMateSnippet snippet) {
    mySnippets.computeIfAbsent(snippet.getKey(), (key) -> Collections.synchronizedList(new ArrayList<>())).add(snippet);
  }

  @Override
  public @NotNull Collection<TextMateSnippet> findSnippet(@NotNull String key, @Nullable TextMateScope scope) {
    if (scope == null) {
      return Collections.emptyList();
    }
    Collection<TextMateSnippet> snippets = mySnippets.get(key);
    if (snippets == null) {
      return Collections.emptyList();
    }
    return new TextMateScopeComparator<>(scope, TextMateSnippet::getScopeSelector).sortAndFilter(snippets);
  }

  @Override
  public @NotNull Collection<TextMateSnippet> getAvailableSnippets(@Nullable TextMateScope scopeSelector) {
    if (scopeSelector == null) {
      return Collections.emptyList();
    }
    return new TextMateScopeComparator<>(scopeSelector, TextMateSnippet::getScopeSelector)
      .sortAndFilter(mySnippets.values().stream().flatMap((values) -> values.stream()));
  }

  public void clear() {
    mySnippets.clear();
  }
}

File: plugins/textmate/core/src/org/jetbrains/plugins/textmate/language/preferences/SnippetsRegistryImpl.kt
package org.jetbrains.plugins.textmate.language.preferences

import kotlinx.collections.immutable.PersistentList
import kotlinx.collections.immutable.persistentListOf
import kotlinx.collections.immutable.persistentMapOf
import org.jetbrains.plugins.textmate.language.TextMateScopeComparatorCore
import org.jetbrains.plugins.textmate.language.syntax.lexer.TextMateScope
import org.jetbrains.plugins.textmate.language.syntax.selector.TextMateSelectorWeigher
import org.jetbrains.plugins.textmate.update
import kotlin.concurrent.atomics.AtomicReference

class SnippetsRegistryBuilder(private val weigher: TextMateSelectorWeigher) {
  private val snippets = AtomicReference(persistentMapOf<String, PersistentList<TextMateSnippet>>())

  fun register(snippet: TextMateSnippet) {
    snippets.update {
      it.put(snippet.key, it[snippet.key]?.add(snippet) ?: persistentListOf(snippet))
    }
  }

  fun build(): SnippetsRegistry {
    return SnippetsRegistryImpl(weigher, snippets.load())
  }
}

class SnippetsRegistryImpl(
  private val weigher: TextMateSelectorWeigher,
  private val snippets: Map<String, List<TextMateSnippet>>,
) : SnippetsRegistry {
  override fun findSnippet(key: String, scope: TextMateScope?): Collection<TextMateSnippet> {
    if (scope == null) {
      return emptyList()
    }
    val snippets = snippets[key] ?: return emptyList()
    return TextMateScopeComparatorCore(weigher, scope, TextMateSnippet::scopeSelector).sortAndFilter(snippets)
  }

  override fun getAvailableSnippets(scopeSelector: TextMateScope?): Collection<TextMateSnippet> {
    if (scopeSelector == null) {
      return emptyList()
    }
    return TextMateScopeComparatorCore(weigher, scopeSelector, TextMateSnippet::scopeSelector)
      .sortAndFilter(snippets.values.flatMap { it })
  }
}
